allensdk/brain_observatory/behavior/project_apis/data_io/behavior_project_cloud_api.py

```diff
@@ -0,0 +1,146 @@
+        self.cache = cache
+        if cache._manifest.metadata_file_names is None:
+            raise RuntimeError("S3CloudCache object has no metadata "
+                               "file names. BehaviorProjectCloudApi "
+                               "expects a S3CloudCache passed which "
+                               "has already run load_manifest()")
+        cache_metadata = set(cache._manifest.metadata_file_names)
+        if cache_metadata != expected_metadata:
+            raise RuntimeError("expected S3CloudCache object to have "
+                               f"metadata file names: {expected_metadata} "
+                               f"but it has {cache_metadata}")
+        self.logger = logging.getLogger("BehaviorProjectCloudApi")
+        self._get_session_table()
+        self._get_behavior_only_session_table()
+        self._get_experiment_table()
+
+    @staticmethod
+    def from_s3_cache(cache_dir: Union[str, Path],
+                      bucket_name: str) -> "BehaviorProjectCloudApi":
+        cache = S3CloudCache(cache_dir, bucket_name)
+        cache.load_latest_manifest()
+        return BehaviorProjectCloudApi(cache)
+
+    def get_behavior_only_session_data(
+            self, behavior_session_id: int) -> BehaviorSession:
+        """get a BehaviorSession by specifying behavior_session_id
+
+        Parameters
+        ----------
+        behavior_session_id: int
+            the id of the behavior_session
+
+        Returns
+        -------
+        BehaviorSession
+
+        """
+        row = self._behavior_only_session_table.query(
+                f"behavior_session_id=={behavior_session_id}")
+        if row.shape[0] != 1:
+            raise RuntimeError("The behavior_only_session_table should have "
+                               "1 and only 1 entry for a given "
+                               "behavior_session_id. For "
+                               f"{behavior_session_id} "
+                               f" there are {row.shape[0]} entries.")
+        data_path = self.cache.download_data(row.data_file_id.values[0])
+        return BehaviorSession.from_nwb_path(str(data_path))
+
+    def get_session_data(self, ophys_session_id: int) -> BehaviorOphysSession:
+        """get a BehaviorOphysSession by specifying session_id
+
+        Parameters
+        ----------
+        ophys_session_id: int
+            the id of the ophys_session
+
+        Returns
+        -------
+        BehaviorOphysSession
+
+        """
+        row = self._session_table.query(
+                f"ophys_session_id=={ophys_session_id}")
+        if row.shape[0] != 1:
+            raise RuntimeError("The behavior_session_table should have "
+                               "1 and only 1 entry for a given "
+                               f"ophys_session_id. For {ophys_session_id} "
+                               f" there are {row.shape[0]} entries.")
+        data_path = self.cache.download_data(row.data_file_id.values[0])
```

Comment:
is the proposed schema in #2007 the planned schema, or has it changed?

Reply:
It is still the planned schema

Reply:
I think, based on this:
https://github.com/AllenInstitute/informatics_data_release_tools/blob/b9d3c5ec8130b8f35824c924ca05b89cacff66ec/src/data_release_tools/core_tools/_00_process_prerelease_metadata.py#L20-L26
that the index won't persist on the round-trip to csv, s3, from_csv.
do I need to set the index here?

Reply:
What I was commenting on was `row.data_file_id`. I don't know that it will be present for ophys_session_table.

Reply:
right, I'll effectively have to join 2 tables, if I udnerstand your point.

Reply:
oh, I see why I am confused. These `ophys_session_id`'s here in the project APIs, for example here:
https://github.com/AllenInstitute/AllenSDK/blob/adea40211e000fd4bae4f644fd0c42f45cae3de4/allensdk/brain_observatory/behavior/project_apis/data_io/behavior_project_lims_api.py#L294
are in fact `ophys_experiment_id`'s:
https://github.com/AllenInstitute/AllenSDK/blob/adea40211e000fd4bae4f644fd0c42f45cae3de4/allensdk/brain_observatory/behavior/session_apis/data_io/behavior_ophys_lims_api.py#L27-L32

Reply:
> 
> 
> ophys_session_table.csv will not have `data_file_id` because ophys nwb files are at the experiment level. Ophys sessions can have multiple experiments. So if the `data_file_id` were stored in this file, it would have to be a list. I think the decision was made to only store ophys experiment nwb files in the "ophys_experiment_table.csv". @njmei correct me if I'm wrong if `data_file_id` will somehow get added through upload to the ophys sessions table.
> 
> I believe the `BehaviorOphysSession` needs to unfortunately be instantiated from `self._experiment_table` given that design.

There is no plan for the data upload tool to modify the `ophys_sessions_table.csv` in any way.